### PR TITLE
Adding AspectChangeIO class from maxjlibs

### PR DESCRIPTION
**Pull request by stepeh**
This version is a lot neater (in terms of interface) than the maxjlibs version.

### CODE REVIEW

File: src/maxpower/kernel/io/AspectChangeIO.maxj [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+	 * @param name The name of the input port (how it is referred to in the manager).
+	 * @param inputType The type of the input data.
+	 * @return The input data you asked for.
+	 */
+	public <T extends KernelObject<T>> T input(String name, KernelType<T> inputType) {
+		return input(name, inputType, constant.var(true));
+	}
+
+	/**
+	 * This method creates an output just like io.input, but does the required aspect changes inside the kernel.
```
Comment (itaygr) [inline src/maxpower/kernel/io/AspectChangeIO.maxj]: should be: "input just like io.input"

File: src/maxpower/kernel/io/AspectChangeIO.maxj [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+		public final DFEVar data;
+		public final DFEVar enable;
+		public final DFEVar extraData;
+		public final DFEVar extraEnable;
+		protected DataWithEnable(DFEVar data, DFEVar enable) {
+			this.data = data;
+			this.enable = enable;
+			extraData = null;
+			extraEnable = null;
+        }
```
Comment (itaygr) [inline src/maxpower/kernel/io/AspectChangeIO.maxj]: indentation.

**Comment by itaygr** [comment]
Can you also support non-blocking inputs?

**Comment by stepeh** [comment]
Hmm... possibly. I'll have a think about it, but I think it shouldn't be
too hard.

**Comment by stepeh** [comment]
OK, this is a lot harder than I initially thought. The problem is that we
have a counter controlling the input, but when we have invalid data then we
want the input to control the counter.

If we take the simple example of a 2:1 aspect change, then on a normal
input the input would be enabled every other cycle. The counter would count
up to 2, enabling the input when the counter is 0, and also controlling the
mux used for the aspect change. However, if we get non-valid data on one
cycle, we don't want to disable the input on the next cycle, as it may have
valid data, so we want to stop the counter.

I can't see any way to fix this right now, but we can discuss it in more
detail next week. Possibly we need to do this inside the manager state
machine instead of in the kernel, but that sounds like a lot of work.

**Comment by chr1sj0nes** [comment]
We can always add non-blocking mode at a later date. I'm happy with the code as is.

**Pull request merged**